src/protocols/bdx/BDXMessageUtils.h

```diff
@@ -0,0 +1,231 @@
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines structures and utility methods for working with BDX
+ *      messages.
+ */
+
+#pragma once
+
+#include <support/CodeUtils.h>
+#include <system/SystemPacketBuffer.h>
+
+namespace chip {
+namespace BDX {
+
+enum ControlMode : uint8_t
+{
+    kNotSpecified  = 0x00,
+    kSenderDrive   = 0x10,
+    kReceiverDrive = 0x20,
+    kAsync         = 0x40,
+};
+
+/*
+ * A structure for representing a SendInit or ReceiveInit message (both contain
+ * identical parameters).
+ */
+struct TransferInit
+{
+    /**
+     * @brief
+     *  Pack (write) the message into a PacketBuffer.
+     *
+     * @param[out] aBuffer a PacketBuffer to use to write the message
+     *
+     * @return CHIP_ERROR Any error that occurs when trying to write to the PacketBuffer
+     */
+    CHIP_ERROR Pack(System::PacketBuffer & aBuffer);
+
+    /**
+     * @brief
+     *  Parse data from an PacketBuffer into a struct instance
+     *
+     * @param[in] aBuffer Pointer to a PacketBuffer containing the data.
+     * @param[out] aParsedMessage Reference to a struct instance where results will be stored.
+     *             Note that this struct will store pointers into the passed PacketBuffer, so it
+     *             is essential that the PacketBuffer is not modified or freed until after the
+     *             TransferInit struct is no longer needed.
+     *
+     * @return CHIP_ERROR Any error that occurs when trying to read the message
+     */
+    static CHIP_ERROR Parse(System::PacketBuffer & aBuffer, TransferInit & aParsedMessage);
+
+    /**
+     * @brief
+     *  Get the size of the message once it is written to a buffer.
+     */
+    size_t PackedSize();
+
+    /**
+     * @brief
+     *  Equality check method.
+     */
+    bool operator==(const TransferInit &) const;
+
+    // Proposed Transfer Control (required)
+    bool mSupportsAsync;
+    bool mSupportsReceiverDrive;
+    bool mSupportsSenderDrive;
+    uint8_t mSupportedVersions;
+
+    // Range Control (required)
+    bool mWideRange;       ///< Set true to indicate mStartOffset and mDefLen are 64-bit, else 32-bit.
+    uint64_t mStartOffset; ///< Proposed start offset of data. 0 for no offset. Will be truncated if mWideRange is false.
+    bool mDefLen;          ///< True if transfer has a definite length.
+
+    uint64_t mMaxLength;    ///< Proposed max length of data in transfer, 0 for indefinite (required if mDefLen is set).
+                            ///< Will be truncated if mWideRange is false.
+    uint16_t mMaxBlockSize; ///< Proposed max block size to use in transfer (required)
+
+    // File designator (required)
+    uint8_t * mFileDesignator;
```

Comment:
These could be pw:span.

What is the lifetime/ownership of these?

Reply:
The expectation for these structs is that the main BDX logic class will use them to write some existing data to a PacketBuffer for sending, or to read from an incoming PacketBuffer and then copy the data elsewhere. I've added some comments explaining that the struct should not outlive the PacketBuffer, but I can add comments here as well. I am also planning on incorporating #2707 once it is fixed.